src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,250 @@
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/apply.h>
+#include <xrpld/app/tx/detail/Batch.h>
+#include <xrpld/ledger/Sandbox.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+XRPAmount
+Batch::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    // Calculate the Inner Txn Fees
+    XRPAmount txnFees{0};
+
+    if (tx.isFieldPresent(sfRawTransactions))
+    {
+        XRPAmount txFees{0};
+        auto const& txns = tx.getFieldArray(sfRawTransactions);
+        for (STObject txn : txns)
+        {
+            STTx const stx = STTx{std::move(txn)};
+            txFees += ripple::calculateBaseFee(view, stx);
+        }
+        txnFees += txFees;
+    }
+
+    // Calculate the BatchSigners Fees
+    std::int32_t signerCount = tx.isFieldPresent(sfBatchSigners)
+        ? tx.getFieldArray(sfBatchSigners).size()
+        : 0;
+
+    return ((signerCount + 2) * view.fees().base) + txnFees;
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const outerAccount = ctx.tx.getAccountID(sfAccount);
+    auto const flags = ctx.tx.getFlags();
+
+    if (flags & tfBatchMask)
+    {
+        JLOG(ctx.j.trace()) << "Batch: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    if (std::popcount(
+            flags &
+            (tfAllOrNothing | tfOnlyOne | tfUntilFailure | tfIndependent)) != 1)
+    {
+        JLOG(ctx.j.trace()) << "Batch: too many flags.";
+        return temMALFORMED;
+    }
+
+    auto const& txns = ctx.tx.getFieldArray(sfRawTransactions);
+
+    if (txns.size() == 0)
+    {
+        JLOG(ctx.j.trace()) << "Batch: txns array is empty.";
+        return temARRAY_EMPTY;
+    }
+
+    if (txns.size() > maxBatchTxCount)
+    {
+        JLOG(ctx.j.trace()) << "Batch: txns array exceeds 8 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    auto const& hashes = ctx.tx.getFieldV256(sfTransactionIDs);
+
+    if (hashes.size() != txns.size())
+    {
+        JLOG(ctx.j.trace()) << "Batch: hashes array size does not match txns.";
+        return temMALFORMED;
+    }
+
+    if (auto const ret = preflight2(ctx); !isTesSuccess(ret))
+        return ret;
```

Comment:
Same here

Reply:
Actually, should be after iterating through the inner transactions, but before checking the `BatchSigners` signatures.